resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -829,50 +968,167 @@ void add(DnsRecord r) {
+                    cache(serverName, authoritativeCache, parent.executor());
+                    return;
+                }
+                serverName = serverName.next;
             }
         }
 
-        // Just walk the linked-list and mark the entry as removed when matched, so next lookup will need to process
-        // one node less.
-        AuthoritativeNameServer remove(String nsName) {
+        // Now handle all AuthoritativeNameServer for which we had no ADDITIONAL record
+        void handleWithoutAdditionals(
+                DnsNameResolver parent, DnsCache cache, AuthoritativeDnsServerCache authoritativeCache) {
             AuthoritativeNameServer serverName = head;
 
             while (serverName != null) {
-                if (!serverName.removed && serverName.nsName.equalsIgnoreCase(nsName)) {
-                    serverName.removed = true;
-                    return serverName;
+                if (serverName.address == null) {
+                    // These will be resolved on the fly if needed.
+                    cacheUnresolved(serverName, authoritativeCache, parent.executor());
+
+                    // Try to resolve via cache as we had no ADDITIONAL entry for the server.
+
+                    List<? extends DnsCacheEntry> entries = cache.get(serverName.nsName, null);
+                    if (entries != null && !entries.isEmpty()) {
+                        InetAddress address = entries.get(0).address();
+
+                        // If address is null we have a resolution failure cached so just use an unresolved address.
+                        if (address != null) {
+                            serverName.update(parent.newRedirectServerAddress(address));
+
+                            for (int i = 1; i < entries.size(); i++) {
+                                address = entries.get(i).address();
+
+                                assert address != null :
+                                        "Cache returned a cached failure, should never return anything else";
+
+                                AuthoritativeNameServer server = new AuthoritativeNameServer(serverName);
+                                server.next = serverName.next;
+                                serverName.next = server;
+                                serverName = server;
+                                serverName.update(parent.newRedirectServerAddress(address));
+
+                                servers++;
+                            }
+                        }
+                    }
                 }
                 serverName = serverName.next;
             }
-            return null;
         }
 
-        int size() {
-            return count;
+        private static void cacheUnresolved(
+                AuthoritativeNameServer server, AuthoritativeDnsServerCache authoritativeCache, EventLoop loop) {
+            // As we have no idea about how long the TTL should be we will just cache the unresolved
```

Comment:
remove this comment? TTL is not used anywhere in this method and this appears to be left over from previous refactoring.